analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/DurableEntityInterfaceRestrictions.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class DurableEntityInterfaceRestrictions : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S6424";
+        private const string MessageFormat = "Use valid entity interface. {0} {1}.";
+        private const string SignalEntityName = "SignalEntity";
+        private const string SignalEntityAsyncName = "SignalEntityAsync";
+        private const string CreateEntityProxyName = "CreateEntityProxy";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var name = (GenericNameSyntax)c.Node;
+                    if (name.Identifier.ValueText is SignalEntityName or SignalEntityAsyncName or CreateEntityProxyName
+                        && name.TypeArgumentList.Arguments.Count == 1
+                        && c.SemanticModel.GetSymbolInfo(name).Symbol is IMethodSymbol method
+                        && IsRestrictedMethod(method)
+                        && method.TypeArguments.Single() is INamedTypeSymbol { TypeKind: not TypeKind.Error } entityInterface
+                        && InterfaceErrorMessage(entityInterface) is { } message)
```

Comment:
It seems we never test whether `entityInterface` is actually an interface. Add a test with a class. I assume the runtime uses [Castle DynamicProxy](http://www.castleproject.org/projects/dynamicproxy/) behind the scenes to generate the proxy and I'm not sure what it does if you pass a concrete class. I think we should raise a warning if you pass a class. Also add a test if a type argument is passed:
```cs
void AzureFunction<T>() => context.SignalEntity<T>(..);
```

Reply:
Good catch! It will also throw => I've added a support.